default/AI/ResourcesAI.py

```diff
@@ -211,30 +376,225 @@ def assess_protection_focus(pid):
+                   local_fleet_rating, regional_threat, sys_status['regional_fleet_threats'])
     return use_protection
 
 
+def use_planet_growth_specials(focus_manager):
+    """set resource foci of planets with potentially useful growth factors. Remove planets from list of candidates."""
+    if useGrowth:
+        # TODO: also consider potential future benefit re currently unpopulated planets
+        for metab, metabIncPop in ColonisationAI.empire_metabolisms.items():
+            for special in [aspec for aspec in AIDependencies.metabolismBoostMap.get(metab, []) if aspec in ColonisationAI.available_growth_specials]:
+                rankedPlanets = []
+                for pid in ColonisationAI.available_growth_specials[special]:
+                    pinfo = focus_manager.all_planet_info[pid]
+                    planet = pinfo.planet
+                    pop = planet.currentMeterValue(fo.meterType.population)
+                    if (pop > metabIncPop - 2 * planet.size) or (GFocus not in planet.availableFoci):  # not enough benefit to lose local production, or can't put growth focus here
+                        continue
+                    for special2 in ["COMPUTRONIUM_SPECIAL"]:
+                        if special2 in planet.specials:
+                            break
+                    else:  # didn't have any specials that would override interest in growth special
+                        print "Considering Growth Focus for %s (%d) with special %s; planet has pop %.1f and %s metabolism incremental pop is %.1f" % (
+                            planet.name, pid, special, pop, metab, metabIncPop)
+                        if pinfo.current_focus == GFocus:
+                            pop -= 4  # discourage changing current focus to minimize focus-changing penalties
+                            rankedPlanets.append((pop, pid, pinfo.current_focus))
+                if not rankedPlanets:
+                    continue
+                rankedPlanets.sort()
+                print "Considering Growth Focus choice for special %s; possible planet pop, id pairs are %s" % (metab, rankedPlanets)
+                for spSize, spPID, cur_focus in rankedPlanets:  # index 0 should be able to set focus, but just in case...
+                    planet = focus_manager.all_planet_info[spPID].planet
+                    if focus_manager.bake_future_focus(spPID, GFocus):
+                        print "%s focus of planet %s (%d) at Growth Focus" % (["set", "left"][cur_focus == GFocus], planet.name, spPID)
+                        break
+                    else:
+                        print "failed setting focus of planet %s (%d) at Growth Focus; focus left at %s" % (planet.name, spPID, planet.focus)
+
+
+def use_planet_production_and_research_specials(focus_manager):
+    """Use production and research specials as appropriate.  Remove planets from list of candidates."""
```

Comment:
Please be more specific in the docstring.

What is happening in the function? Does it set any foci? Does it only calculate them? How does it "use" specials.